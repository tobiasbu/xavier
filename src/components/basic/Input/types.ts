import * as React from 'react';

/**
 * Describe the input text properties.
 */
export interface InputProps {
  /**
   * Input label
   * @default 'Standard'
   */
  label?: string;
  /**
   * Default input value
   * @default ''
   */
  value?: string | number;
  /**
   * Is the input disabled?
   * @default false
   */
  disabled?: boolean;
  /**
   * Is this input validated?
   * 
   * - true: validated
   * - false: invalid
   * - null: none
   * @default null
   */
  validation?: boolean | null;
  /**
   * Placeholder text
   * @default null
   */
  placeholder?: string;
  /**
   * Reference of the real input DOM element.
   */
  forwardedRef?: (el : HTMLInputElement) => void;
  /**
   * On input.
   */
  onInput?: React.ChangeEventHandler<HTMLInputElement>;
  /**
   * On change.
   */
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
  /**
   * Is this input required for form validation?
   */  
  required?: boolean;
  /**
   * Default value of this input.
   */
  defaultValue?: React.ReactText;
  /**
   * Input error message.
   */
  errorMessage?: React.ReactText;
};