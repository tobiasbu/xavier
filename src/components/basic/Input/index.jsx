import React, { useState } from 'react';
import PropTypes from 'prop-types';
import slugify from 'slugify';

import useCallbackRef from '@utils/useCallbackRef';
import * as Utils from '@utils';
import * as commons from '../commons';

import useStyle from './style';

/**
 * @typedef { import("./types").InputProps } InputProps
 */

/**
 * Astro standard input.
 *
 * I should probably create a component...
 *
 * @param {InputProps} props Properties
 */
const Input = (props) => {
  const {
    label, value, disabled, required, validation, placeholder, forwardedRef,
    errorMessage, onInput, onChange,
  } = props;

  const classes = useStyle();
  const [isLabelFloating, setLabelFloat] = useState(commons.shouldLabelFloat(props));
  const [elementRef, setRef] = useCallbackRef();
  const disabledClass = (disabled) ? ' a-input--disabled' : '';
  const validationClass = commons.getValidationClass(validation);

  const inputId = Utils.generateHash(props, 'm-i-');
  const labelId = Utils.generateHash(props, 'm-l-');

  const onKeyUp = () => {
    const r = commons.shouldLabelFloat(props, elementRef);
    if (isLabelFloating === r) {
      return;
    }
    setLabelFloat(r);
  };

  return (
    <div className={`a-input${disabledClass}${validationClass} ${classes.inputContainer}`}>
      <input
        type="text"
        defaultValue={value}
        placeholder={placeholder}
        disabled={disabled}
        onInput={onInput}
        onChange={onChange}
        onKeyUp={onKeyUp}
        id={inputId}
        aria-labelledby={labelId}
        required={required}
        aria-required={required ? 'true' : 'false'}
        name={label ? slugify(label, { lower: true }) : ''}
        ref={(e) => {
          if (forwardedRef) {
            forwardedRef(e);
          }
          setRef(e);
        }}
      />
      <label
        className={isLabelFloating ? 'a-input__label--floating' : ''}
        id={labelId}
        htmlFor={inputId}
      >
        {label}
      </label>
      <span className={`a-input__error ${classes.errorMessage}`}>{errorMessage}</span>
    </div>
  );
};

/**
 * Input default props
 */
Input.defaultProps = {
  label: 'Standard',
  value: '',
  disabled: false,
  required: false,
  validation: null,
  errorMessage: '',
  placeholder: null,
  forwardedRef: null,
  onInput: null,
  onChange: null,
};

/**
 * Input prop types
 */
Input.propTypes = {
  label: PropTypes.string,
  value: PropTypes.string,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  validation: PropTypes.bool,
  errorMessage: PropTypes.string,
  placeholder: PropTypes.string,
  onInput: PropTypes.func,
  onChange: PropTypes.func,
  forwardedRef: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({ current: PropTypes.instanceOf(HTMLInputElement) }),
  ]),
};

export default Input;
