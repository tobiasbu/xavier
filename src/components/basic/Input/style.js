import { createUseStyles } from 'react-jss';

export default createUseStyles({
  inputContainer: {
    margin: '1rem',
  },
});
