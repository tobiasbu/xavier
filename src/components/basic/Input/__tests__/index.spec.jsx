import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import renderer from 'react-test-renderer';

import Input from '..';

describe('Input', () => {
  it('should render', () => {
    const component = shallow(<Input label="Descrição" />);
    expect(toJSON(component)).toMatchSnapshot();
  });

  it('check value', () => {
    const component = renderer.create(<Input label="Dinheiro, pra que, Dinheiro?" defaultValue="Se ela não me da bola..." />);
    const intance = component.root;
    const input = intance.findByType('input');
    expect(input.props.defaultValue).toBe('Se ela não me da bola...');
  });
});
