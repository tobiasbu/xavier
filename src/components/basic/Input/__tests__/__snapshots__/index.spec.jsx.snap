// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render 1`] = `
<div
  className="a-input inputContainer-0-2-1"
>
  <input
    aria-labelledby="descricao-label"
    aria-required="false"
    defaultValue=""
    disabled={false}
    id="descricao-input"
    name="descricao"
    onChange={null}
    onInput={null}
    onKeyUp={[Function]}
    placeholder={null}
    required={false}
    type="text"
  />
  <label
    className=""
    htmlFor="descricao-input"
    id="descricao-label"
  >
    Descrição
  </label>
  <span
    aria-live="polite"
    className="a-input__error errorMessage-0-2-2"
  />
</div>
`;
